lib/membrane_mp4/muxer.ex

```diff
@@ -0,0 +1,149 @@
+defmodule Membrane.MP4.Muxer do
+  @moduledoc """
+  Puts payloaded streams into an MPEG-4 container.
+  """
+
+  # Due to the structure of MPEG-4 containers, it is not possible
+  # to send buffers with mdat box content right after they are
+  # processed - we need to know the size of this box up front.
+  # The current solution requires storing all incoming buffers
+  # in memory until the last track ends.
+  # Once some kind of seeking mechanism is implemented, it will
+  # be possible to send chunks of samples on demand, only
+  # correcting the mdat header after all media data was send.
+
+  use Membrane.Filter
+
+  alias Membrane.MP4.Container
+  alias __MODULE__.{MovieBox, Track}
+
+  @ftyp [
+          ftyp: %{
+            children: [],
+            fields: %{
+              compatible_brands: ["isom", "iso2", "avc1", "mp41"],
+              major_brand: "isom",
+              major_brand_version: 512
+            }
+          }
+        ]
+        |> Container.serialize!()
+
+  @mdat_data_offset 8
+
+  @default_samples_per_chunk 10
+
+  def_input_pad :input,
+    demand_unit: :buffers,
+    caps: Membrane.MP4.Payload,
+    availability: :on_request
+
+  def_output_pad :output, caps: :buffers
+
+  def_options tracks: [
+                spec: pos_integer,
+                default: 1,
+                description: "Number of tracks that the muxer should expect"
+              ],
+              samples_per_chunk: [
```

Comment:
Actually it is not related to the type of a track. It may happen that we receive all buffers for some tracks before even  receiving caps for some others. In that situation, we need to know the total number of tracks up front to be able to decide whether we should wait or proceed.

I think about changing the default value to `2`, though - it seems to be a more common usage pattern.

Reply:
My comment was on line 48 - `samples_per_chunk`